src/providers/token-properties-provider.ts

```diff
@@ -0,0 +1,110 @@
+import { BigNumber } from '@ethersproject/bignumber';
+import { ChainId, Token } from '@uniswap/sdk-core';
+
+import { log } from '../util';
+
+import { ICache } from './cache';
+import { ProviderConfig } from './provider';
+import { ITokenFeeFetcher, TokenFeeResult } from './token-fee-fetcher';
+import { DEFAULT_ALLOWLIST } from './token-validator-provider';
+
+const DEFAULT_TOKEN_BUY_FEE_BPS = BigNumber.from(0);
+const DEFAULT_TOKEN_SELL_FEE_BPS = BigNumber.from(0);
+
+// on detector failure, assume no fee
+const DEFAULT_TOKEN_FEE_RESULT = {
+  buyFeeBps: DEFAULT_TOKEN_BUY_FEE_BPS,
+  sellFeeBps: DEFAULT_TOKEN_SELL_FEE_BPS,
+};
+
+const DEFAULT_TOKEN_PROPERTIES_RESULT: TokenPropertiesResult = {
+  tokenFeeResult: DEFAULT_TOKEN_FEE_RESULT,
+};
+
+type Address = string;
+export type TokenPropertiesResult = {
+  tokenFeeResult: TokenFeeResult;
+};
+export type TokenPropertiesMap = Record<Address, TokenPropertiesResult>;
+
+export interface ITokenPropertiesProvider {
+  getTokensProperties(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenPropertiesMap>;
+}
+
+export class TokenPropertiesProvider implements ITokenPropertiesProvider {
+  private CACHE_KEY = (chainId: ChainId, address: string) =>
+    `token-properties-${chainId}-${address}`;
+
+  constructor(
+    private chainId: ChainId,
+    private tokenPropertiesCache: ICache<TokenPropertiesResult>,
+    private tokenFeeFetcher: ITokenFeeFetcher,
+    private allowList = DEFAULT_ALLOWLIST
+  ) {}
+
+  public async getTokensProperties(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenPropertiesMap> {
+    const tokenToResult: TokenPropertiesMap = {};
+    const addressesToFetchFeesOnchain: string[] = [];
+    const addressesRaw = this.buildAddressesRaw(tokens);
+
+    const tokenProperties = await this.tokenPropertiesCache.batchGet(addressesRaw)
+
+    // Check if we have cached token validation results for any tokens.
+    for (const address of addressesRaw) {
+      const cachedValue = tokenProperties[address];
+      if (cachedValue) {
+        tokenToResult[address] = cachedValue;
+      } else if (this.allowList.has(address)) {
```

Comment:
I made it such that `TokenPropertiesProvider` calls out to `TokenValidatorProvider` first. As part of it, if the token is in the allowlist, we hardcode that token to UNKNOWN result type, so that the allowlist tokens will not call the `TokenFeeFetcher` since they are not FOT.